src/Lean/PrettyPrinter/Delaborator/Builtins.lean

```diff
@@ -247,12 +267,37 @@ def delabAppImplicit : Delab := do
       pure (fnStx, paramKinds.tailD [], argStxs))
   let stx := Syntax.mkApp fnStx argStxs
 
-  if ← isRegularApp then
+  if ← isRegularApp maxArgs then
     (guard (← getPPOption getPPNotation) *> unexpandRegularApp stx)
     <|> (guard (← getPPOption getPPStructureInstances) *> unexpandStructureInstance stx)
     <|> pure stx
   else pure stx
 
+/--
+Delaborates applications. Removes up to `maxArgs` arguments to form
+the "head" of the application and delaborates the head using `delabHead`.
+The remaining arguments are processed depending on whether heuristics indicate that the application
+should be delaborated using `@`.
+-/
+def delabAppCore (maxArgs : Nat) (delabHead : Delab) : Delab := do
```

Comment:
Are there any delabs in core that should already be using `delabAppCore` with a specific `maxArgs`?

Reply:
Yeah, many builtin delaborators could use it. One example is `delabProjectionApp`, which could use it to handle over-application.

My plan was to use `delabAppCore` to upstream a more general version of [`Lean.PrettyPrinter.Delaborator.withOverApp`](https://github.com/leanprover/std4/blob/9dd24a3493cceefa2bede383f21e4ef548990b68//Std/Lean/Delaborator.lean#L24-L38) and then use this combinator to improve builtins.

Reply:
I went ahead and upstreamed the combinator and improved the builtins that could be improved.